File: pre_post/ComputerScience/Algorithm/LeetCode/239. Sliding Window Maximum.py
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        if nums is None:
            return []
        #elemens in window represent the position in nums
        window,result=[],[]
        
        for i,key in enumerate(nums):
            
            #window is not empty, but full  
            if window and window[0]<=i-k:
                window.pop(0) 
                
            #window is not empty, some elements is smaller than key
            while(window and nums[window[-1]]<key):
                window.pop()
            
            window.append(i)
            if i>=k-1:
                result.append(nums[window[0]])
        
        return result 

File: pre_post/ComputerScience/Algorithm/LeetCode/239. Sliding Window Maximum.cpp
class Solution {
 public:
  vector<int> maxSlidingWindow(vector<int>& nums, int k) {
    if (nums.empty()) return vector<int>();
    deque<int> window;
    vector<int> res;
    for (int i = 0; i < nums.size(); i++) {
      if (!window.empty() && window.front() == i - k) window.pop_front();
      // 把 window 中所有小于新加入元素的值都干掉
      while (!window.empty() && nums[window.back()] < nums[i]) {
        window.pop_back();
      }
      window.push_back(i);
      if (i >= k - 1) res.push_back(nums[window.front()]);
    }
    return res;
  }
};